example_plugins/keyboard/src/lib.rs

```diff
@@ -0,0 +1,197 @@
+        }
+        if self.s_is_pressed {
+            move_vector += Vector3::new(-1., 0., 0.)
+        }
+        if self.d_is_pressed {
+            move_vector += Vector3::new(0., 0., 1.)
+        }
+
+        // Send movement command to server
+        if move_vector != Vector3::zeros() {
+            let distance = move_vector.normalize() * frame_time.delta * 10.;
+
+            let command = MoveCommand { distance };
+
+            io.send(&command);
+        }
+    }
+}
+
+impl UserState for ServerState {
+    fn new(io: &mut EngineIo, sched: &mut EngineSchedule<Self>) -> Self {
+        // Define how the cube should be rendered
+
+        // Create one cube entity at the origin, and make it synchronize to clients
+        let cube_entity = io.create_entity();
+        io.add_component(cube_entity, &Transform::default());
+        io.add_component(
+            cube_entity,
+            &Render::new(CUBE_HANDLE).primitive(Primitive::Triangles),
+        );
+        io.add_component(cube_entity, &Synchronized);
+        io.add_component(cube_entity, &CubeFlag);
+
+        // Create the Update system, which interprets movement commands and updates the transform
+        // component on the object with CubeFlag
+        sched.add_system(
+            Self::update,
+            SystemDescriptor::new(Stage::Update)
+                .subscribe::<MoveCommand>()
+                .query::<CubeFlag>(Access::Write)
+                .query::<Transform>(Access::Write),
+        );
+
+        Self
+    }
+}
+
+impl ServerState {
+    fn update(&mut self, io: &mut EngineIo, query: &mut QueryResult) {
+        // Check for movement commands
+        if let Some(MoveCommand { distance }) = io.inbox_first() {
+            // Update each object accordingly
+            for key in query.iter() {
+                query.modify::<Transform>(key, |tf| {
+                    tf.pos += distance;
+                })
+            }
+        }
+    }
+}
+
+/// Defines the mesh data fro a cube
+fn cube() -> Mesh {
+    let size = 0.25;
+    let vertices = vec![
+        Vertex::new([-size, -size, -size], [0.0, 1.0, 1.0]),
+        Vertex::new([size, -size, -size], [1.0, 0.0, 1.0]),
+        Vertex::new([size, size, -size], [1.0, 1.0, 0.0]),
+        Vertex::new([-size, size, -size], [0.0, 1.0, 1.0]),
+        Vertex::new([-size, -size, size], [1.0, 0.0, 1.0]),
+        Vertex::new([size, -size, size], [1.0, 1.0, 0.0]),
+        Vertex::new([size, size, size], [0.0, 1.0, 1.0]),
+        Vertex::new([-size, size, size], [1.0, 0.0, 1.0]),
+    ];
+
+    let indices = vec![
+        3, 1, 0, 2, 1, 3, 2, 5, 1, 6, 5, 2, 6, 4, 5, 7, 4, 6, 7, 0, 4, 3, 0, 7, 7, 2, 3, 6, 2, 7,
+        0, 5, 4, 1, 5, 0,
+    ];
+
+    Mesh { vertices, indices }
+}
+
+impl Component for CubeFlag {
+    const ID: ComponentIdStatic = ComponentIdStatic {
+        id: pkg_namespace!("Cube Flag"),
+        size: 0,
+    };
+}
+
+impl Message for MoveCommand {
```

Comment:
Whats this impl for? Is it documented in The Book?

Reply:
More or less, the "Core Concepts" page talks about what information is contained in a message, and the ECS example also uses this impl and has comments explaining it. We should definitely put a chapter in the book though